File: findMinStops.cpp
#include <bits/stdc++.h>
using namespace std;

#define MOD 1000000007
#define INF 10000000000000007
#define MAX 1000000
 
void dijkstra(vector<pair<int,int>>v[] , int start , int size)
{
	vector<int>d(MAX , INF);
 
	d[start] = 0;
	set<pair<int,int>>st;
	st.insert({0 , start});
	
	vector<int>parent(size + 1);
	parent[1] = -1;
 
	while(!st.empty())
	{
        int node = st.begin()->second;
        st.erase(st.begin());
        
        if(node == size)
        break;
 
        //go to adjacent elements of v.
        for(auto edge: v[node])
        {
        	int to = edge.first;
        	int wt = edge.second;
        	if(d[to] > d[node] + wt)
        	{
        		st.erase({d[to] , to});
        		d[to] = d[node] + wt;
        		parent[to] = node;
        		st.insert({d[to] , to});
 
        	}
        }
 
	}
 
  if(d[size] == INF)
  {
   cout << "IMPOSSIBLE";
    return;
  }
 
  vector<int>path;
 
  
  int k = size;
  while(k != -1){
      path.push_back(k);
      k = parent[k];
  }
 reverse(path.begin() , path.end());
 cout << path.size();
 cout << endl;
  
}
 
int main()
{
   
 
	int n , m;
	cin >> n >> m;
	vector<pair<int,int>>graph[MAX];
	int u , v;
	for(int i=0;i<m;i++)
	{
      cin >> u >> v;
       graph[u].push_back({v,1});
       graph[v].push_back({u,1});
	}
 
	//call dijkstra from 1 ..
 
	dijkstra(graph , 1 , n);
 
}

File: findMinStops.java
import sys
import heapq

INF = 10000000000000007
MAX = 1000000

def dijkstra(graph, start, size):
    d = [INF] * MAX
    d[start] = 0

    st = []
    heapq.heappush(st, (0, start))

    parent = [-1] * (size + 1)
    parent[1] = -1

    while st:
        dist, node = heapq.heappop(st)

        if node == size:
            break

        for to, wt in graph[node]:
            if d[to] > d[node] + wt:
                d[to] = d[node] + wt
                parent[to] = node
                heapq.heappush(st, (d[to], to))

    if d[size] == INF:
        print("IMPOSSIBLE")
        return

    path = []
    k = size
    while k != -1:
        path.append(k)
        k = parent[k]

    path.reverse()
    print(len(path))
   

def main():
    input = sys.stdin.read
    data = input().split()
    index = 0

    n = int(data[index])
    index += 1
    m = int(data[index])
    index += 1

    graph = [[] for _ in range(MAX)]

    for _ in range(m):
        u = int(data[index])
        index += 1
        v = int(data[index])
        index += 1
        graph[u].append((v, 1))
        graph[v].append((u, 1))

    dijkstra(graph, 1, n)

if __name__ == "__main__":
    main()
